### PR TITLE
Export huffman loopkup init table

### PR DESCRIPTION
**Pull request by nicolas-grekas**
Right now, huffman lookup init takes 150ms on my machine on the cli (85ms with opcache).

This table has been generated with the patch below applied to Symfony VarExporter's Exporter class, then this just before returning in HuffmanLookupInit:
```php
ksort($encodingAccess);
foreach ($encodingAccess as $k => $v) {
    ksort($encodingAccess[$k], SORT_STRING);
}
file_put_contents(
    __DIR__.'/huffman-lookup.php',
    "<?php\n\nreturn ".\Symfony\Component\VarExporter\VarExporter::export($encodingAccess).";\n"
);
```

<details>

```diff
--- a/src/Symfony/Component/VarExporter/Internal/Exporter.php
+++ b/src/Symfony/Component/VarExporter/Internal/Exporter.php
@@ -212,6 +212,8 @@ class Exporter
         $subIndent = $indent.'    ';
 
         if (\is_string($value)) {
+            return '"'.str_replace('%', '\x', rawurlencode($value)).'"';
+
             $code = sprintf("'%s'", addcslashes($value, "'\\"));
 
             $code = preg_replace_callback('/([\0\r\n]++)(.)/', function ($m) use ($subIndent) {
@@ -240,6 +242,10 @@ class Exporter
         }
 
         if (\is_array($value)) {
+            if ([0, 1] === array_keys($value) && !\is_array($value[0]) && !\is_array($value[1])) {
+                return '['.self::export($value[0]).', '.self::export($value[1]).']';
+            }
+
             $j = -1;
             $code = '';
             foreach ($value as $k => $v) {
```

</details>

### CODE REVIEW

**Comment by kelunik** [comment]
We're definitely interested in optimizing the lookup table. Time spent can be significant, but also the 33 MB RAM are rather huge, especially when compared to the nghttp2 implementation.

I haven't really looked into the lookup table construction, but maybe we can optimize that, too, maybe even with a decreased decoding performance in favor of a better init performance. 

Maybe we should also base the used variant on the SAPI being used and whether opcache is available.

**Comment by kelunik** [comment]
As I just saw it on your other profile, did you have xdebug loaded while benchmarking?

**Comment by trowski** [comment]
Yes, I'm surprised that RAM usage increased for CLI, but decreased for webserver, or am I misunderstanding something?

**Comment by nicolas-grekas** [comment]
RAM usage increases because PHP needs memory to parse the new big file.
It decreases on the web because it leverages opcache's shared memory better I suppose.
But I know how to improve this even further, update coming :)

**Comment by nicolas-grekas** [comment]
Et voilà the new benchmark:
image

I confirm xdebug was disabled.

The right side is all zeros because the call was so fast that it has been pruned by Blackfire.

The reason for the memory going to zero is that the table is now a purely static array, which means it stays in opcache's shared memory only.

PR ready :)

File: src/Internal/HPackNative.php [hunk 1]
```diff
@@ -214,7 +219,7 @@ public static function huffmanDecode(string $input) /* : ?string */
         for ($bitCount = $off = $i = 0; $i < $length; $i++) {
             $currentByte = $input[$i];
 
-            [$lookup, $chr] = $lookup[$currentByte];
+            [$lookup, $chr] = (\is_int($lookup) ? $huffmanLookup[$lookup] : $lookup)[$currentByte];
```
Comment (nikic) [inline src/Internal/HPackNative.php]: Might make more sense to just always use the int variant?

Comment (nicolas-grekas) [reply src/Internal/HPackNative.php]: That'd require much more changes to the map generation logic. Not sure I can do it.

Comment (nicolas-grekas) [reply src/Internal/HPackNative.php]: OK, I did it :)

**Comment by nicolas-grekas** [comment]
Note that we could compact each line in `huffman-lookup.php` into single strings, replacing the myriad of small arrays.
But now that these arrays are in shared memory, this would provide nothing more.

PR ready.

**Comment by kelunik** [comment]
Thanks! We can look into further optimizations in another PR if it's necessary.

**Pull request merged**